samples/estimation/df-chemistry/chemistry.py

```diff
@@ -0,0 +1,482 @@
+        assert len(two_body_result.one_norms) == \
+               len(two_body_result.two_norms) \
+               and len(two_body_result.one_norms) == \
+               len(two_body_result.eigenvalues) \
+               and len(two_body_result.one_norms) == \
+               len(two_body_result.eigenvectors)
+
+
+# Convert 2D array into string representation
+def ndarray2d_to_string(arr):
+    str_arr = []
+    for elem in arr:
+        str_arr.append(np.array2string(elem, separator=','))
+    return f"[{','.join(str_arr)}]"
+
+
+# The script takes one required positional argument, URI of the FCIDUMP file
+parser = ArgumentParser(description='Double-factorized chemistry sample')
+# Use n2-10e-8o as the default sample.
+# Pass a different filename to get estimates for different compounds
+parser.add_argument('-f', '--fcidumpfile',
+                    default='https://aka.ms/fcidump/n2-10e-8o',
+                    help='Path to the FCIDUMP file describing the Hamiltonian')
+args = parser.parse_args()
+
+# ----- Read the FCIDUMP file and get resource estimates from Q# algorithm -----
+structure = FCIDumpFileContent.from_file(args.fcidumpfile)
+df = DoubleFactorization.process_fcidump(structure, 0.001)
+
+# Load Q# project
+qsharp.init(project_root=".")
+
+# Construct the Q# operation call for which we need to perform resource estimate
+str_one_body_eigenvalues = np.array2string(df.one_body_eigenvalues,
+                                           separator=',')
+
+str_one_body_eigenvectors = ndarray2d_to_string(df.one_body_eigenvectors)
+
+str_two_body_eigenvalues = ndarray2d_to_string(df.two_body_eigenvalues)
+
+str_two_body_eigenvectors = "[" + \
+    ','.join([ndarray2d_to_string(eigenvectors)
+              for eigenvectors in df.two_body_eigenvectors]) + "]"
+
+qsharp_string = \
+    "Microsoft.Quantum.Applications.Chemistry.DoubleFactorizedChemistry(" + \
+    "Microsoft.Quantum.Applications.Chemistry.DoubleFactorizedChemistryProblem(" + \
+    f"{df.num_orbitals}, {df.one_body_norm}, {df.two_body_norm}, " + \
+    f"{str_one_body_eigenvalues}, {str_one_body_eigenvectors}, " + \
+    f"[1.0, size = {df.rank}], {str_two_body_eigenvalues}, " + \
+    f"{str_two_body_eigenvectors})," + \
+    "Microsoft.Quantum.Applications.Chemistry.DoubleFactorizedChemistryParameters(0.001,))"
+
+# Get resource estimates
+res = qsharp.estimate(qsharp_string,
+                      params={"errorBudget": 0.01,
+                              "qubitParams": {"name": "qubit_maj_ns_e6"},
+                              "qecScheme": {"name": "floquet_code"}})
+
+# Store estimates in json file
+with open('resource_estimate.json', 'w') as f:
+    f.write(res.json)
+
+# Print high-level resource estimation results
+print(f"Algorithm runtime: {res['physicalCountsFormatted']['runtime']}")
+print(f"Number of physical qubits required: {res['physicalCountsFormatted']['physicalQubits']}")
+print("For more detailed resource counts, see file resource_estimate.json")
```

Comment:
Any reason we made this a straight Python file rather than a notebook where we could use our widgets to visualize the final results?

Reply:
When I think about working on a large project with a fair amount of Python and Q# code and reading from/writing to files, I'm more likely to do that with plain files than with a notebook - I feel that Jupyter Notebooks target smaller amounts of code and emphasize visualization more than other aspects. (We worked on an arithmetic RE project in the past, and it was a lot easier to follow what happened in the Python file than in a Python notebook.) Plus, we don't have any examples of multi-file Q# projects with Python code that is not in a notebook, and it's good to have a variety of examples of different ways one can structure they Q# project.

With this code already written, we can next improve the sample to add a notebook "frontend" that would load the existing Python and Q# from the "backend" and focus on showing how to provide different parameters to RE, like Mathias suggested, and producing visualizations. But I would like to publish this core functionality of the sample first, before we look into further improvements.